virttest/qemu_devices/qdevices.py

```diff
@@ -1667,62 +1763,75 @@ def _start_daemon(self):
-        logging.info("Created virtiofs daemon process with parent PID %d.",
-                     self._daemon_process.get_pid())
-
-    def _stop_daemon(self):
-        """Stop the virtiofs daemon."""
-        if self._daemon_process is not None:
-            try:
-                if not utils_misc.wait_for(lambda: not self._daemon_alive(), 3):
-                    raise DeviceError('The virtiofs daemon is still alive.')
-            finally:
-                self._daemon_process.close()
-                self._daemon_process = None
+        self.set_param('cmd', fsd_cmd)
+        self.set_param('run_bg_kwargs', {'output_func': self._handle_log,
+                                         'auto_close': False})
+        self.set_param('status_active', 'Waiting for vhost-user socket connection')
+        super(QVirtioFSDev, self).start_daemon()
 
-    def clear(self):
-        """Clear the virtiofs daemon."""
-        try:
-            self._stop_daemon()
-        except DeviceError:
-            pass
+    def __eq__(self, other, dynamic=True):
+        if super(QVirtioFSDev, self).__eq__(other, dynamic):
+            return self.get_param('sock_path') == other.get_param('sock_path')
+        return False
 
-    def hotplug(self, monitor):
-        self._start_daemon()
+    def __ne__(self, other):
+        return super(QVirtioFSDev, self).__ne__(other)
 
-    def verify_hotplug(self, out, monitor):
-        return self._daemon_alive()
 
-    def unplug(self, monitor):
-        self._stop_daemon()
+class QswtpmDev(QDaemonDev):
+    """
+    Virtual swtpm pseudo device.
+    """
+    def __init__(self, aobject, binary, sock_path, storage_path,
+                 version=None, extra_options=None):
+        """
+        :param aobject: The auto object of swtpm daemon.
+        :type aobject: str
+        :param binary: The binary of swtpm daemon.
+        :type binary: str
+        :param sock_path: The socket path of of swtpm daemon.
+        :type sock_path: str
+        :param storage_path: The path to the directory for tpm state
+        :type storage_path: str
+        :param version: the version of tpm functionality.
+        :type version: str
+        :type storage_path: str
+        :param extra_options: The external options of swtpm daemon.
+        :type extra_options: str
+        """
+        super(QswtpmDev, self).__init__('vtpm', aobject=aobject,
+                                        child_bus=QUnixSocketBus(sock_path, aobject))
+        self.set_param('binary', binary)
+        self.set_param('sock_path', sock_path)
+        self.set_param('storage_path', storage_path)
+        self.set_param('version', version)
+        self.set_param('extra_options', extra_options)
 
-    def verify_unplug(self, out, monitor):
-        """Verify the status of virtiofs daemon."""
-        return not self._daemon_alive()
+    def start_daemon(self):
+        tpm_cmd = '%s socket --daemon' % self.get_param('binary')
```

Comment:
Should check the existence of `/usr/bin/swtpm` somewhere?

Reply:
Hi @vivianQizhu 

I think we should not check the existence of binary since the user should ensure it and install the corresponding binary before testing if wanna run this part. The user should know such a background of TPM feature, should know to provide the binary.Like qmeu binary, if you wanna run avocado-vt, you should install the qemu binary and other environments settings.

The binary path is decided by the product, it's hardly changed in the long term, I think. We could update the path at that time if it is changed really. How about?

And if not found the binary, the VM will run error and stop then we could check the error info by the log file. Check the existence just raise the error in advance, but there is no actual sense since this error is involved by environment setting issue, not a product bug or automation bug.

BTW, not check the existence of `/usr/libexec/virtiofsd` in `def fs_define_by_params(self, name, params, bus=None):` ethier.
But up to you, if you wanna add, will add.
Thanks.

Reply:
@YongxueHong Okay, let it be.

Reply:
> @YongxueHong Okay, let it be.

Hi @vivianQizhu 
For which? check the existence?
Thanks.

Reply:
Hi, @vivianQizhu 
Have updated, please help me to review.
Thanks.

Reply:
> @YongxueHong Okay, let it be.

Sorry for not saying it clearly, I meant to follow your own idea, keep your original code...well then...it's fine anyway.

Reply:
> > @YongxueHong Okay, let it be.
> 
> Sorry for not saying it clearly, I meant to follow your own idea, keep your original code...well then...it's fine anyway.

Hi, @vivianQizhu 
Thanks for your understanding, I prefer the original code, I will revert it.
Thanks.